base_layer/wallet/src/ffi.rs

```diff
@@ -29,87 +29,402 @@
 //! will be exposed via FFI and will consist of API calls that the FFI client can make into the Wallet module and a set
 //! of Callbacks that the client must implement and provide to the Wallet module to receive asynchronous replies and
 //! updates.
-
-use crate::output_manager_service::service::PendingTransactionOutputs;
+extern crate libc;
+use crate::{output_manager_service::service::PendingTransactionOutputs, Wallet};
 use chrono::NaiveDateTime;
-use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
-use tari_comms::connection::NetAddress;
+use libc::{c_char, c_int, c_uchar, c_uint, c_ulonglong};
+// use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
+use crate::{
+    output_manager_service::{handle::OutputManagerResponse::TransactionCancelled, OutputManagerConfig},
+    wallet::WalletConfig,
+};
+use std::{
+    boxed::Box,
+    ffi::{CStr, CString},
+    process::Output,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+use tari_comms::{
+    connection::NetAddress,
+    control_service::ControlServiceConfig,
+    peer_manager::{NodeIdentity, Peer, PeerFeature, PeerFeatures, PeerNodeIdentity},
+};
+use tari_comms_dht::envelope::NodeDestination::NodeId;
 use tari_core::{
-    transaction::{Transaction, UnblindedOutput},
+    tari_amount::MicroTari,
+    transaction::{
+        OutputFeatures,
+        OutputFlags,
+        Transaction,
+        TransactionInput,
+        TransactionKernel,
+        TransactionOutput,
+        UnblindedOutput,
+    },
     types::{PrivateKey, PublicKey},
 };
+use tari_crypto::keys::SecretKey;
+use tari_p2p::initialization::CommsConfig;
+use tari_utilities::{hex::Hex, ByteArray};
+use tokio::runtime::Runtime;
 
-// use tari_core::transaction_protocol::sender::RawTransactionInfo;
+pub type TariWallet = Wallet;
+pub type WalletDateTime = NaiveDateTime;
 
-// TODO implement this
-pub struct WalletFfi {}
+/// -------------------------------- ByteVector ------------------------------------------------ ///
+pub struct ByteVector(Vec<c_uchar>); // declared like this so that it can be exposed to external header
 
-pub struct IdentityFfi {
-    public_key: PublicKey,
-    secret_key: PrivateKey,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_create() -> *mut ByteVector {
+    let bytes = ByteVector(Vec::new());
+    Box::into_raw(Box::new(bytes))
 }
 
-pub struct KeyManagerStateFfi {
-    master_seed: PrivateKey,
-    branch_seed: String,
-    index: c_uint,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_destroy(bytes: *mut ByteVector) {
+    if bytes.is_null() {
+        let b = Box::from_raw(bytes);
+    }
+}
+
+/// adds c_char to internal vec
+pub unsafe extern "C" fn bytevector_add_byte(bytes: *mut ByteVector, byte: c_uchar) {
+    (*bytes).0.push(byte);
+}
```

Comment:
Is this not extremely inefficient? To stream a byte stream across one byte at a time. Surely we can pass a fully formed byte array and a length? We know the lengths ahead of time. We can do it with strings why not pure bytes?

Reply:
A huge portion of the data are these byte arrays so it needs to be easy to do and as efficient as possible.

Reply:
Fixed in the next commit.

Method was removed. Now bytevector_create takes an array pointer of c_uchar, and an element count of c_int and constructs the inner vector using a slice::from_raw_parts(*c_uchar,c_int)

Thanks for the assist here.